### PR TITLE
Support python 3.11

### LINKED ISSUE #236: Python 3.11 not supported due to vendored `wrapt`
On a 3.11 system, this error is thrown:

```python
Error processing line 1 of /opt/warehouse/lib/python3.11/site-packages/appmap.pth:

  Traceback (most recent call last):
    File "<frozen site>", line 186, in addpackage
    File "<string>", line 1, in <module>
    File "/opt/warehouse/lib/python3.11/site-packages/appmap/__init__.py", line 2, in <module>
      from _appmap import generation  # noqa: F401
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/opt/warehouse/lib/python3.11/site-packages/_appmap/__init__.py", line 1, in <module>
      from . import configuration
    File "/opt/warehouse/lib/python3.11/site-packages/_appmap/configuration.py", line 16, in <module>
      from _appmap.labels import LabelSet
    File "/opt/warehouse/lib/python3.11/site-packages/_appmap/labels.py", line 5, in <module>
      from .importer import Filterable
    File "/opt/warehouse/lib/python3.11/site-packages/_appmap/importer.py", line 10, in <module>
      from appmap import wrapt
    File "/opt/warehouse/lib/python3.11/site-packages/appmap/wrapt/__init__.py", line 10, in <module>
      from .decorators import (adapter_factory, AdapterFactory, decorator,
    File "/opt/warehouse/lib/python3.11/site-packages/appmap/wrapt/decorators.py", line 34, in <module>
      from inspect import ismethod, isclass, formatargspec
  ImportError: cannot import name 'formatargspec' from 'inspect' (/usr/local/lib/python3.11/inspect.py)
```

This us due to the vendored `wrapt` module version 1.12.1 not supporting Python 3.11 - they introduced that in 1.14.0
See: https://wrapt.readthedocs.io/en/latest/changes.html#version-1-14-0

### PR DESCRIPTION
**Pull request by apotterri**
**Please don't merge this, it needs a rebase**

Fixes #236.

### CODE REVIEW

File: _appmap/unittest.py [hunk 1]
```diff
@@ -12,39 +13,59 @@ def _get_test_location(cls, method_name):
...
                 yield
-        else:
-            # session.record may return None
-            yield
+
+else:
+    # unittest.case._Outcome.testPartExecutor is used by all supported
+    # versions of unittest to run test cases. `isTest` will be True when
+    # the part is the actual test method, False when it's setUp or
+    # teardown.
```
Comment (dividedmind) [inline _appmap/unittest.py:58]: It seems like you copied this comment verbatim and forgot to update it.

File: _appmap/unittest.py [hunk 2]
```diff
@@ -12,39 +13,59 @@ def _get_test_location(cls, method_name):
...
-    location = _get_test_location(test_case.__class__, method_name)
-    with _session.record(
-        test_case.__class__, method_name, location=location
-    ) as metadata:
-        if metadata:
-            with wrapped(
-                *args, **kwargs
-            ), testing_framework.collect_result_metadata(metadata):
+if sys.version_info[1] < 8:
+    # unittest.case._Outcome.testPartExecutor is used by all supported
```
Comment (dividedmind) [inline _appmap/unittest.py:17]: No longer 'all supported', right?

**Review by dividedmind — CHANGES_REQUESTED** [review]
Judging by the CI log, it seems that moving `wrapt` to `_appmap` might not have been the best idea. What is the reason to move, how was it more difficult to upgrade when it was in `appmap`? The commit message doesn't make it entirely clear.

Other than that and the comments, please add 3.11 to CI (and maybe move smoke test to that).

File: _appmap/unittest.py (hunk 1 above)
Comment (apotterri) [reply _appmap/unittest.py:58]: Oops, yeah, thanks. Fixed.

File: _appmap/importer.py [hunk 3]
```diff
@@ -264,12 +262,26 @@ def wrapped_find_spec(find_spec, _, args, kwargs):
...
-        logger.debug("no find_spec for finder %r", finder)
-        return
+    # Prior to 3.11, it worked fine to just grab find_spec from the finder and wrap it. The
+    # implementation of builtin finders must have changed with 3.11, because we now need the same
+    # kind of workaround we use above for exec_module.
+    if sys.version_info[1] < 11:
+        find_spec = getattr(finder, "find_spec", None)
+        if find_spec is None:
+            logger.debug("no find_spec for finder %r", finder)
+            return
```
[the remark this answers is not in the document]
Comment (dividedmind) [reply _appmap/importer.py:272]: Yes, but wouldn't `getattr_static` work universally?

[the remark this answers is not in the document]
Comment (apotterri) [reply _appmap/importer.py:272]: Oops, got the sense of your question, backwards. Sorry about that.

Yeah, `getattr_static` would probably work universally. It's just such a gross thing to have to do.... ;)

I'll check and make sure it works, but assuming it does, I'll just use that.

[the remark this answers is not in the document]
Comment (apotterri) [reply _appmap/importer.py:272]: In fact, it [doesn't work](https://app.travis-ci.com/github/getappmap/appmap-python/jobs/602678808), so I'll just leave this as-is.

**Comment by apotterri** [comment]
@dividedmind this is ready for another look

I made one fairly substantial change since you last looked: I removed the `wrapt` submodule, and replaced it by the use of the `vendoring` package. This seems to be a more [widely-used](https://github.com/pypa/pip/blob/main/pyproject.toml#L29) [approach](https://github.com/python-poetry/poetry-core/blob/5da9a2ed898d9438a8235169f3a4833a1380acdc/pyproject.toml#L117) to [vendoring](https://github.com/search?q=path%3A**%2Fpyproject.toml+tool.vendoring&type=code). One of its benefits is that it will allow us to easily vendor more packages, which we may need sooner rather than later: https://appmap-group.slack.com/archives/C03GM7D2QRH/p1684786853793949?thread_ts=1684779111.885699&cid=C03GM7D2QRH

File: README.md [hunk 4]
```diff
@@ -83,9 +92,9 @@ Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
 
 ```
 
-[Note that the current configuration requires a 10.0 for the Travis build to pass. To make
-this easier to achieve, convention and refactoring checks have both been disabled. They
-should be reenabled as soon as possible.]
+[Note that the current configuration has a threshold set which must be met for the Travis build to
+pass. To make this easier to achieve, a number of checks have both been disabled. They should be
```
Comment (dividedmind) [inline README.md:96]: Good catch :)

**Review by dividedmind — APPROVED** [review]
Looks good, thanks for the changes! (I think I liked submoduling more, but if vendoring is easier to work with, that's fine.)

**Pull request merged**